Add preliminary error paths, minor fixes

jonlamb-gh opened the pull request:
- added various results to the functions that can provide them (only I/O related for now)
- stubbed out an error enum place holder to funnel the various types of HAL errors
- fix the DAC impl, updated HAL to be half-duplex by default (will split the peripheral later on)
- a low quality refactor of main that groups the various module related activities in an effort towards making error handling less awkward
- added an example error handler that ignores CAN Tx timeout errors

jonlamb-gh committed:
Add preliminary error paths, minor fixes

- added various results to the functions that can provide them (only I/O related for now)
- stubbed out an error enum place holder to funnel the various types of HAL errors
- fix the DAC impl, updated HAL to be half-duplex by default (will split the peripheral later on)
- a low quality refactor of main to group the various module related activities to make error handling less awkward for now
- added an example error handler that ignores CAN Tx timeout errors

ZackPierce commented on src/brake/kia_soul_ev_niro/brake_module.rs:
```diff
@@ -84,38 +85,55 @@ impl UnpreparedBrakeModule {
             self.brake_pedal_position.prevent_signal_discontinuity();
 
-            self.brake_dac.output_ab(
+            if let Err(_) = self.brake_dac.output_ab(
```
clippy is most likely going to recommend the use of `.is_err()` instead of the inner-content-dropping `let` expression.

ZackPierce commented on src/brake/kia_soul_ev_niro/brake_module.rs:
```diff
@@ -84,38 +85,55 @@ impl UnpreparedBrakeModule {
 
-    fn enable_control(&mut self, debug_console: &mut DebugConsole) {
+    fn enable_control(&mut self, debug_console: &mut DebugConsole) -> Result<(), OxccError> {
+        let mut result = Ok(());
```
It is usually more Rustic to construct a result at the point of decision/return rather than creating and mutating a result output variable. The fact that if/else branches return values that can in turn be automatically returned from functions if they are the final statement makes this easier.

ZackPierce commented on src/main.rs:
```diff
@@ -201,6 +232,112 @@ fn main() -> ! {
     }
 }
 
+fn check_and_process_control_can_frames(
```
The use of `and` in a method name is usually a negative code smell. I don't see any checking being done, perhaps just `process_control_can_frames` will suffice.

ZackPierce commented on src/main.rs:
```diff
@@ -201,6 +232,112 @@ fn main() -> ! {
+    can_gateway: &mut CanGatewayModule,
+) -> Result<(), OxccError> {
+    // attempt to publish them all, only report the first to fail
+    let mut result = Ok(());
```
Contrary to what the comment says, this pattern will only result in the *last* failure being reported.

ZackPierce commented on src/main.rs:
```diff
@@ -201,6 +232,112 @@ fn main() -> ! {
+
+    // it is typically to get timeout errors if the CAN bus is not active or
+    // there are no other nodes connected to it
+    if error == OxccError::CanBusTxTimeout {
```
As this system becomes more complicated, it will likely become worthwhile to make a more action-local decision to ignore/not-propagate the precursor errors to this variant.  Limiting the spread of locally-remediable has the upside of reducing the temptation to do deeper global state inspection at error-handling time, which in turn can lead to subtle interconnection between seemingly unrelated aspects of state-tracking and error codes.

Another alternative to preventing error-handler tendril creep is to carry more decision-making data with the error objects themselves. e.g. I can imagine a future where certain CAN bus timeouts are completely ignorable, while others require some different action, and that differences in response can be captured or inferable from data in the error variant.

jonlamb-gh replied:
Totally agree with you on that one. I've got that pushed down a little further in the stack now, with notes around it likely changing in the near future.

We definitely ought to do something in most cases.
One easy idea I had was to put the CAN bus in a passive mode, disable controls and ignore further enable commands until we can successfully transmit a frame or error bits are cleared/etc.
I'll run some other ideas by you on implementing parts of the error management once the main logic is in better shape.

ZackPierce commented on src/throttle_module.rs:
```diff
@@ -83,57 +84,76 @@ impl UnpreparedThrottleModule {
                 ranges::coerce(ThrottleSpoofHighSignal::clamp(spoof_command_high)),
                 ranges::coerce(ThrottleSpoofHighSignal::clamp(spoof_command_low)),
-            );
+            )?;
```
If this is an entirely private and internal function, you may consider *not* converting the error type to the top-level/global error type immediately, and instead return the more immediate/domain-specific error type so that the other wrapping functions (still in this module) can make the choice about how to handle it without having to destructure/dig back down to the right variant or otherwise deal with dropped/simplified error state.

jonlamb-gh replied:
Great idea, I'll work on moving fault handling mechanics into the modules :+1:

ZackPierce reviewed (commented):
Added some thoughts and nits.

Another important error path to consider is if any failure that might occur during the `prepare_module` activities is important enough to prevent the modules stepping into a usable state.

jonlamb-gh committed:
Address concerns

- comments/concerns on error path management
- clippy suggestions
- use domain-specific error types, push CAN tx timeouts down further
- a proper safe-from-any-context error handler coming soon (possibly with a panic impl)
- keeping place-holder logic in main, will be refactored as more action-local decisions are implemented

jonlamb-gh committed:
Undo changes to src/ranges.rs
